infinite-tracing/src/main/java/com/newrelic/ResponseObserver.java

```diff
@@ -1,98 +1,117 @@
         this.logger = logger;
-        this.disconnectionHandler = disconnectionHandler;
-        this.shouldRecreateCall = shouldRecreateCall;
+        this.channelManager = channelManager;
+        this.aggregator = aggregator;
+        this.backoffManager = backoffPolicy;
     }
 
     @Override
     public void onNext(V1.RecordStatus value) {
-        metricAggregator.incrementCounter("Supportability/InfiniteTracing/Response");
+        aggregator.incrementCounter("Supportability/InfiniteTracing/Response");
+        backoffManager.reset();
     }
 
     @Override
     public void onError(Throwable t) {
-        if (isChannelClosing(t)) {
-            logger.log(Level.FINE, "Stopping current gRPC call because the channel is closing.");
-            return;
-        }
+        Status status = Status.fromThrowable(t);
 
-        if (isAlpnError(t)) {
-            logger.log(Level.SEVERE, t, "ALPN does not appear to be supported on this JVM. Please install a supported JCE provider or update Java to use Infinite Tracing");
-            metricAggregator.incrementCounter("Supportability/InfiniteTracing/NoALPNSupport");
-            disconnectionHandler.terminate();
+        // If the span observer is knowingly cancelled, it is canceled with a ChannelClosingException,
+        // which is detected here and ignored.
+        if (status.getCause() instanceof ChannelClosingException) {
+            logger.log(Level.FINE, "Stopping gRPC response observer because span observer was closed by another thread.");
             return;
         }
 
-        if (!isConnectionTimeoutException(t)) {
-            logger.log(Level.WARNING, t, "Encountered gRPC exception");
+        if (isOkHttpALPNError(status)) {
+            logger.log(Level.SEVERE, t,
+                    "ALPN does not appear to be supported on this JVM. Please install a supported JCE provider or update Java to use Infinite Tracing.");
+            aggregator.incrementCounter("Supportability/InfiniteTracing/NoALPNSupport");
+            channelManager.shutdownChannelForever();
+            return;
         }
 
-        metricAggregator.incrementCounter("Supportability/InfiniteTracing/Response/Error");
+        aggregator.incrementCounter("Supportability/InfiniteTracing/Span/gRPC/" + status.getCode());
+        aggregator.incrementCounter("Supportability/InfiniteTracing/Response/Error");
 
-        Status status = null;
-        if (t instanceof StatusRuntimeException) {
-            StatusRuntimeException statusRuntimeException = (StatusRuntimeException) t;
-            status = statusRuntimeException.getStatus();
-            metricAggregator.incrementCounter("Supportability/InfiniteTracing/Span/gRPC/" + status.getCode());
+        if (status.getCode() == Status.Code.UNIMPLEMENTED) {
+            // See: https://source.datanerd.us/agents/agent-specs/blob/master/Infinite-Tracing.md#unimplemented
```

Comment:
ugg I wish these specs were public

Reply:
> Has this been manually tested and verified to work (old and new functionality alike)? And have the AITs been run against this branch? 

Yes - lots of manual testing, and the AITs have passed.